src/StreamJsonRpc/Reflection/MessageFormatterProgressTracker.cs

```diff
@@ -0,0 +1,168 @@
+            }
+
+            lock (this.ProgressLock)
+            {
+                long progressId = this.NextProgressId++;
+                this.RequestProgressMap.Add(this.RequestIdBeingSerialized.Value, progressId);
+
+                this.ProgressMap.Add(progressId, new ProgressParamInformation(value));
+
+                return progressId;
+            }
+        }
+
+        /// <summary>
+        /// Class used to keep relevant information of an object that implements <see cref="IProgress{T}"/>.
+        /// </summary>
+        public class ProgressParamInformation
+        {
+            /// <summary>
+            /// Initializes a new instance of the <see cref="ProgressParamInformation"/> class.
+            /// </summary>
+            /// <param name="progressObject">The object implementing <see cref="IProgress{T}"/>.</param>
+            public ProgressParamInformation(object progressObject)
+            {
+                Requires.NotNull(progressObject, nameof(progressObject));
+
+                Type iProgressOfTType = MessageFormatterProgressTracker.FindIProgressOfT(progressObject.GetType());
+
+                Verify.Operation(iProgressOfTType != null, Resources.FindIProgressOfTError);
+
+                this.ValueType = iProgressOfTType.GenericTypeArguments[0];
+                this.ReportMethod = iProgressOfTType.GetRuntimeMethod(nameof(IProgress<int>.Report), new Type[] { this.ValueType });
+                this.ProgressObject = progressObject;
+            }
+
+            /// <summary>
+            /// Gets the actual <see cref="Type"/> reported by <see cref="IProgress{T}"/>.
+            /// </summary>
+            public Type ValueType { get; }
+
+            /// <summary>
+            /// Gets the <see cref="MethodInfo"/> of <see cref="IProgress{T}.Report(T)"/>.
+            /// </summary>
+            public MethodInfo ReportMethod { get; }
+
+            /// <summary>
+            /// Gets the instance of the object implementing <see cref="IProgress{T}"/>.
+            /// </summary>
+            public object ProgressObject { get; }
+        }
+
+        /// <summary>
+        /// Class that implements <see cref="IProgress{T}"/> and sends <see cref="ProgressRequestSpecialMethod"/> notification when reporting.
+        /// </summary>
+        public class JsonProgress<T> : IProgress<T>
+        {
+            private readonly JsonRpc rpc;
+            private readonly long? token;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="JsonProgress{T}"/> class.
+            /// </summary>
+            /// <param name="rpc">The <see cref="JsonRpc"/> instance used to send the <see cref="ProgressRequestSpecialMethod"/> norification.</param>
+            /// <param name="token">The <see cref="long"/> progress token used to obtain the <see cref="ProgressParamInformation"/> intance from <see cref="ProgressMap"/>.</param>
+            public JsonProgress(JsonRpc rpc, long? token)
```

Comment:
>long? [](start = 45, length = 5)

We can't assume that the client is using numbers for their progress tokens. This can hopefully be simply `object` like in the JsonRpcRequest class. The formatter may provide a `JToken` or `long`, or whatever the formatter will be able to re-serialize when transmitting the message.
We need to add at least a couple `JsonRpcServerInteropTests` to the mix to exercise both number and non-number progress tokens to make sure we've shaken out all the `long` assumptions from the code.

Reply:
I thought the client would only provide the Progress object and then we assign a token to it and replace it in the message, storing the Progress object in the dictionary. I made the change for object anyway since I think it won't affect any functionality.

---
In reply to: [314177878](https://github.com/microsoft/vs-streamjsonrpc/pull/320#discussion_r314177878) [](ancestors = 314177878)

Reply:
Yes, on the *client* side you're right. But this `JsonProgress<T>` class isn't used on the client at all. It's used on the server-side. And on the server side, we can't control what the token will be since the client determines the token.

---
In reply to: [314497199](https://github.com/microsoft/vs-streamjsonrpc/pull/320#discussion_r314497199) [](ancestors = 314497199,314177878)

Reply:
Leaving this active till we get server interop tests to validate we work properly with non-long tokens.

---
In reply to: [314849437](https://github.com/microsoft/vs-streamjsonrpc/pull/320#discussion_r314849437) [](ancestors = 314849437,314497199,314177878)